Task 3/Task 3.2/Task3.2.1.Library/DynamicArray.cs

```diff
@@ -0,0 +1,281 @@
+
+            return false;
+        }
+
+        /// <summary>
+        /// Метод Insert, позволяющий добавить элемент в произвольную позицию массива (обратите
+        /// внимание, может потребоваться расширить массив). Метод должен возвращать true, если
+        /// добавление прошло успешно и false в противном случае.При выходе за границу массива
+        /// должно генерироваться исключение ArgumentOutOfRangeException.
+        /// </summary>
+        public bool Insert(int index, T item) // 7
+        {
+            if (index < 0 || index > Length)
+                throw new ArgumentOutOfRangeException();
+
+            RightShift(index, 1);
+            _array[index] = item;
+            return true;
+        }
+
+
+        /// <summary>
+        /// Добавить метод ToArray, возвращающий новый массив (обычный), содержащий все
+        /// содержащиеся в текущем динамическом массиве объекты.
+        /// </summary>
+        public T[] ToArray() // 4*
+        {
+            T[] newArray = new T[Length];
+
+            Array.Copy(_array, newArray, Length);
+
+            return newArray;
+        }
+
+
+        /// <summary>
+        /// Метод, реализующий интерфейс IEnumerable.
+        /// </summary>
+        IEnumerator IEnumerable.GetEnumerator() => (IEnumerator)GetEnumerator(); // 10
+
+        /// <summary>
+        /// Метод, реализующий интерфейс IEnumerable<T>.
+        /// </summary>
+        public IEnumerator<T> GetEnumerator() => new DynamicArrayEnum<T>(this); // 10
+
+        /// <summary>
+        /// Реализовать интерфейс ICloneable для создания копии массива.
+        /// </summary>
+        public object Clone() // 3*
+        {
+            DynamicArray<T> dyClone = new DynamicArray<T>(this);
+            dyClone.SetCapacity(Capacity);
+
+            return dyClone;
+        }
+
+        /// <summary>
+        /// Индексатор, позволяющий работать с элементом с указанным номером. При выходе за
+        /// границу массива должно генерироваться исключение ArgumentOutOfRangeException.
+        /// * Доступ к элементам с конца при использовании отрицательного индекса (−1: последний,
+        /// −2: предпоследний и т.д.).
+        /// </summary>
+        public T this[int index] // 11
+        {
+            get
+            {
+                index = IndexHelper(index);
+
+                return _array[index];
+            }
+            set
+            {
+                index = IndexHelper(index);
+
+                _array[index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Check and return correct index
+        /// </summary>
+        private int IndexHelper(int index)
+        {
+            if (index >= Length || index < 0 - Length)
+                throw new ArgumentOutOfRangeException();
+
+            if (index >= 0)
+            {
+                return index;
+            }
+            else
+            {
+                return index + Length;
+            }
+        } //1*
```

Comment:
`IndexHelper` is a variable/class name. For method you should use verbs and try to describe what it actually does. For example `ResolveIndex()`